test/integration/udproute_test.go

```diff
@@ -356,6 +356,45 @@ func TestUDPRouteEssentials(t *testing.T) {
 		_, err := resolver.LookupHost(ctx, "kernel.org")
 		return err != nil
 	}, ingressWait, waitTick)
+
+	t.Log("testing port matching")
```

Comment:
Now wait a minute, I got it [from you](https://github.com/Kong/kubernetes-ingress-controller/blame/main/test/integration/httproute_test.go#L282-L284)!

I'm not sure what you mean by subtest, but the typical means of splitting these is to create an entirely separate test function. This works, but it requires a lot of boilerplate, since a large portion of the test is just setting up the environment, e.g. 
[udp_new_test.diff.txt](https://github.com/Kong/kubernetes-ingress-controller/files/11150924/udp_new_test.diff.txt)

Of approximately 600 lines, only 20 are performing the actual test. The rest are setting up the Gateway(Class), UDPRoute, and test Deployment, and are identical to the Essentials test. If we find that we need to change that code, we need to remember to do it in two places.

The L4 tests furthermore have an additional complication: we can only run one at a time due to the single shared test port, so that additional boilerplate setup eats up more execution time than most.

Copying alone seems bad on that basis; I would lean towards just accepting the additional 20-30 lines on the existing test. We could look at factoring out shared components into helpers, similar to [this prior work](https://github.com/Kong/kubernetes-ingress-controller/issues/2461), but designing those well takes care to make them flexible and expandable. The [Ingress helpers](https://github.com/Kong/kubernetes-testing-framework/blob/238e12e74d32e8692fcae1435b45fd8490f70ac9/pkg/utils/kubernetes/generators/ingress.go#L27-L60) have caused me trouble in the past by being quite prescriptive and making assumptions that don't work for all tests (who would ever need more than a single rule!), doubly so since they're in KTF and require a separate package release.

I think making a common reusable base environment (namely the Gateway(Class) and test Service) that we can keep running under a set of smaller tests makes sense, but that refactor is out of scope for the goal of making UDPRoute do port matching, so I'll create a separate issue for it.

Reply:
> Now wait a minute, I got it [from you](https://github.com/Kong/kubernetes-ingress-controller/blame/main/test/integration/httproute_test.go#L282-L284)!

😅 

What I meant here is to do `t.Run("testing port matching", func(t *testing.T){...` to isolate it both logically and visually.